File: ios-app/Doompedia/Ranking/CardKeywords.swift
import Foundation

enum CardKeywords {
    private static let stableTopics: Set<String> = [
        "science",
        "technology",
        "history",
        "geography",
        "culture",
        "politics",
        "economics",
        "sports",
        "health",
        "environment",
        "society",
        "biography",
        "general",
    ]

    private static let keywordBuckets: [(String, [String])] = [
        ("biography", ["born", "died", "actor", "author", "scientist", "politician", "player"]),
        ("science", ["physics", "chemistry", "biology", "mathematics", "astronomy", "scientific"]),
        ("technology", ["software", "computer", "internet", "digital", "algorithm", "device"]),
        ("history", ["war", "empire", "century", "historical", "revolution", "ancient", "dynasty"]),
        ("geography", ["city", "country", "region", "river", "mountain", "capital", "province"]),
        ("politics", ["government", "election", "parliament", "policy", "minister", "president"]),
        ("culture", ["music", "film", "literature", "art", "religion", "language"]),
        ("economics", ["economy", "finance", "market", "trade", "industry", "currency"]),
        ("health", ["medicine", "disease", "medical", "health", "hospital", "symptom"]),
        ("sports", ["football", "basketball", "olympic", "athlete", "league", "championship"]),
        ("environment", ["climate", "ecology", "forest", "wildlife", "pollution", "conservation"]),
    ]

    private static let stopwords: Set<String> = [
        "about", "after", "before", "their", "there", "which", "while", "where", "these", "those",
        "through", "using", "under", "between", "during", "known", "wikipedia", "article", "entry",
        "first", "second", "third", "world", "state", "city", "country",
    ]

    static func preferenceKeys(
        title: String,
        summary: String,
        topicKey: String,
        maxKeys: Int = 12
    ) -> [String] {
        let text = "\(title) \(summary)".lowercased()
        var ordered: [String] = []
        var seen = Set<String>()

        func append(_ value: String) {
            let cleaned = value.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !cleaned.isEmpty, !seen.contains(cleaned) else { return }
            seen.insert(cleaned)
            ordered.append(cleaned)
        }

        let canonicalPrimary = canonicalTopic(topicKey)
        if !canonicalPrimary.isEmpty, canonicalPrimary != "general" {
            append(canonicalPrimary)
        }

        for (topic, keywords) in keywordBuckets where keywords.contains(where: { text.contains($0) }) {
            append(topic)
        }

        if title.range(of: "list of", options: .caseInsensitive) != nil { append("culture") }
        if title.range(of: "university", options: .caseInsensitive) != nil { append("society") }
        if title.range(of: "city", options: .caseInsensitive) != nil { append("geography") }
        if title.range(of: "war", options: .caseInsensitive) != nil { append("history") }

        if ordered.isEmpty {
            append("general")
        }

        for token in tokenize(text: text) {
            if token.count >= 4, !stopwords.contains(token) {
                append(token.replacingOccurrences(of: " ", with: "-"))
            }
            if ordered.count >= maxKeys { break }
        }

        return Array(ordered.prefix(maxKeys))
    }

    static func primaryTopic(title: String, summary: String, topicKey: String) -> String {
        let keys = preferenceKeys(title: title, summary: summary, topicKey: topicKey, maxKeys: 8)
        if let stable = keys.first(where: { stableTopics.contains($0) && $0 != "general" }) {
            return stable
        }
        let fallback = canonicalTopic(topicKey)
        return fallback.isEmpty ? "general" : fallback
    }

    static func displayTags(
        title: String,
        summary: String,
        topicKey: String,
        bookmarked: Bool,
        maxTags: Int = 6
    ) -> [String] {
        var ordered: [String] = []
        var seen = Set<String>()

        func append(_ value: String) {
            guard !value.isEmpty, !seen.contains(value) else { return }
            seen.insert(value)
            ordered.append(value)
        }

        for key in preferenceKeys(title: title, summary: summary, topicKey: topicKey, maxKeys: maxTags * 2) where stableTopics.contains(key) {
            append(prettyTopic(key))
        }

        if title.range(of: "list of", options: .caseInsensitive) != nil { append("Lists") }
        if title.range(of: "university", options: .caseInsensitive) != nil { append("Education") }
        if bookmarked { append("Saved") }

        return Array(ordered.prefix(maxTags))
    }

    static func prettyTopic(_ raw: String) -> String {
        let normalized = raw
            .replacingOccurrences(of: "-", with: " ")
            .replacingOccurrences(of: "_", with: " ")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty else { return "General" }
        return normalized
            .split(separator: " ")
            .map { token in
                let lower = token.lowercased()
                return lower.prefix(1).uppercased() + lower.dropFirst()
            }
            .joined(separator: " ")
    }

    private static func canonicalTopic(_ rawTopic: String) -> String {
        let canonical = rawTopic
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .replacingOccurrences(of: "_", with: "-")
            .replacingOccurrences(of: " ", with: "-")
            .replacingOccurrences(of: "-+", with: "-", options: .regularExpression)

        switch canonical {
        case "history-of":
            return "history"
        case "geography-of":
            return "geography"
        case "economy-of":
            return "economics"
        case "list-of":
            return "culture"
        default:
            return canonical
        }
    }

    private static func tokenize(text: String) -> [String] {
        let pattern = "[A-Za-z][A-Za-z-]{2,}"
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return [] }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        return regex.matches(in: text, options: [], range: range).compactMap { match in
            guard let tokenRange = Range(match.range, in: text) else { return nil }
            return String(text[tokenRange]).lowercased()
        }
    }
}

File: android-app/app/src/main/java/com/feelbachelor/doompedia/domain/CardKeywords.kt
package com.feelbachelor.doompedia.domain

object CardKeywords {
    private val stableTopics = setOf(
        "science",
        "technology",
        "history",
        "geography",
        "culture",
        "politics",
        "economics",
        "sports",
        "health",
        "environment",
        "society",
        "biography",
        "general",
    )

    private val keywordBuckets: List<Pair<String, List<String>>> = listOf(
        "biography" to listOf("born", "died", "actor", "author", "scientist", "politician", "player"),
        "science" to listOf("physics", "chemistry", "biology", "mathematics", "astronomy", "scientific"),
        "technology" to listOf("software", "computer", "internet", "digital", "algorithm", "device"),
        "history" to listOf("war", "empire", "century", "historical", "revolution", "ancient", "dynasty"),
        "geography" to listOf("city", "country", "region", "river", "mountain", "capital", "province"),
        "politics" to listOf("government", "election", "parliament", "policy", "minister", "president"),
        "culture" to listOf("music", "film", "literature", "art", "religion", "language"),
        "economics" to listOf("economy", "finance", "market", "trade", "industry", "currency"),
        "health" to listOf("medicine", "disease", "medical", "health", "hospital", "symptom"),
        "sports" to listOf("football", "basketball", "olympic", "athlete", "league", "championship"),
        "environment" to listOf("climate", "ecology", "forest", "wildlife", "pollution", "conservation"),
    )

    private val stopwords = setOf(
        "about", "after", "before", "their", "there", "which", "while", "where", "these", "those",
        "through", "using", "under", "between", "during", "known", "wikipedia", "article", "entry",
        "first", "second", "third", "world", "state", "city", "country",
    )

    private val tokenRegex = Regex("[A-Za-z][A-Za-z-]{2,}")

    fun preferenceKeys(
        title: String,
        summary: String,
        topicKey: String,
        maxKeys: Int = 12,
    ): List<String> {
        val text = "$title $summary".lowercase()
        val ordered = linkedSetOf<String>()

        val canonicalPrimary = canonicalTopic(topicKey)
        if (canonicalPrimary.isNotBlank() && canonicalPrimary != "general") {
            ordered += canonicalPrimary
        }

        keywordBuckets.forEach { (topic, keywords) ->
            if (keywords.any { keyword -> text.contains(keyword) }) {
                ordered += topic
            }
        }

        if (title.contains("list of", ignoreCase = true)) ordered += "culture"
        if (title.contains("university", ignoreCase = true)) ordered += "society"
        if (title.contains("city", ignoreCase = true)) ordered += "geography"
        if (title.contains("war", ignoreCase = true)) ordered += "history"

        if (ordered.isEmpty()) {
            ordered += "general"
        }

        tokenRegex.findAll(text).forEach { match ->
            val token = match.value.lowercase()
            if (token.length >= 4 && token !in stopwords) {
                ordered += token.replace(' ', '-')
            }
            if (ordered.size >= maxKeys) return@forEach
        }

        return ordered.take(maxKeys)
    }

    fun primaryTopic(title: String, summary: String, topicKey: String): String {
        val keys = preferenceKeys(title = title, summary = summary, topicKey = topicKey, maxKeys = 8)
        return keys.firstOrNull { it in stableTopics && it != "general" }
            ?: canonicalTopic(topicKey).ifBlank { "general" }
    }

    fun displayTags(
        title: String,
        summary: String,
        topicKey: String,
        bookmarked: Boolean,
        maxTags: Int = 6,
    ): List<String> {
        val ordered = linkedSetOf<String>()
        preferenceKeys(
            title = title,
            summary = summary,
            topicKey = topicKey,
            maxKeys = maxTags * 2,
        ).forEach { key ->
            if (key in stableTopics) {
                ordered += prettyTopic(key)
            }
        }

        if (title.contains("list of", ignoreCase = true)) ordered += "Lists"
        if (title.contains("university", ignoreCase = true)) ordered += "Education"
        if (bookmarked) ordered += "Saved"

        return ordered.take(maxTags)
    }

    fun prettyTopic(raw: String): String {
        return raw
            .replace('-', ' ')
            .replace('_', ' ')
            .trim()
            .split(' ')
            .filter { it.isNotBlank() }
            .joinToString(" ") { token -> token.lowercase().replaceFirstChar { it.titlecase() } }
            .ifBlank { "General" }
    }

    private fun canonicalTopic(rawTopic: String): String {
        val canonical = rawTopic
            .trim()
            .lowercase()
            .replace('_', '-')
            .replace(' ', '-')
            .replace(Regex("-+"), "-")
        return when (canonical) {
            "history-of" -> "history"
            "geography-of" -> "geography"
            "economy-of" -> "economics"
            "list-of" -> "culture"
            else -> canonical
        }
    }
}

